### ISSUE TITLE
BPF Program too large on Cilium 1.15.0-rc.0 / Talos 1.6.1 / Kernel 6.1.69

### ISSUE THREAD

**Issue by ekarlso** [cilium/cilium#30093]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

Upgraded cilium to see if it would help on issues mentioned on slack and here https://github.com/cilium/cilium/issues/28753. Then I get error with too large bpf program.

### Cilium Version

cilium image (running): 1.15.0-rc.0
[cilium.log](https://github.com/cilium/cilium/files/13823368/cilium.log)

### Kernel Version

Talos (v1.6.1)   6.1.69-talos     containerd://1.7.11

### Kubernetes Version

1.28.5

### Anything else?

```
---
k8s:
  requireIPv4PodCIDR: true
  requireIPv6PodCIDR: false

bpf:
  masquerade: true

enableIPv4BIGTCP: false
enableIPv4Masquerade: true

enableIPv6BIGTCP: false
enableIPv6Masquerade: false

localRedirectPolicy: true

ipam:
  operator:
    clusterPoolIPv4PodCIDRList:
      - 172.17.16.0/20
    clusterPoolIPv4MaskSize: 26

    clusterPoolIPv6PodCIDRList: []
    clusterPoolIPv6MaskSize: 128
    autoCreateCiliumPodIPPools: {}

ipv4:
  enabled: true
ipv6:
  enabled: false

hostPort:
  enabled: true

nodePort:
  enabled: true

enableIdentityMark: true
remoteNodeIdentity: true
wellKnownIdentities:
  enabled: true

enableCnpStatusUpdates: true
endpointStatus:
  enabled: true
  status: policy health controllers log state

devices:
  - eth0

hostFirewall:
  enabled: true # enable host policies
extraConfig:
  allow-localhost: policy # enable policies for local host

hubble:
  enabled: true
  metrics:
    enableOpenMetrics: true

    enabled:
      - drop
      - tcp
      - flow
      - flows-to-world
      - icmp
      - httpV2
      - port-distribution

  relay:
    enabled: true
    rollOutPods: true

    prometheus:
      enabled: true

    securityContext:
      seccompProfile:
        type: RuntimeDefault
      allowPrivilegeEscalation: false
      runAsNonRoot: true
      capabilities:
        drop:
          - ALL
  ui:
    enabled: true
    rolloutPods: true

    backend:
      securityContext:
        seccompProfile:
          type: RuntimeDefault
        allowPrivilegeEscalation: false
        runAsNonRoot: true
        runAsUser: 65535
        capabilities:
          drop:
            - ALL

    frontend:
      securityContext:
        seccompProfile:
          type: RuntimeDefault
        allowPrivilegeEscalation: false
        runAsNonRoot: true
        runAsUser: 65535
        capabilities:
          drop:
            - ALL

prometheus:
  enabled: true

tunnel: vxlan

routingMode: tunnel

kubeProxyReplacement: true
k8sServiceHost: 127.0.0.1
k8sServicePort: 7445

ingressController:
  enabled: false
  default: true

gatewayAPI:
  enabled: true

l2announcements:
  enabled: true

rolloutCiliumPods: true

operator:
  replicas: 2
  rolloutPods: true

  prometheus:
    enabled: false

  securityContext:
    seccompProfile:
      type: RuntimeDefault
    allowPrivilegeEscalation: false
    capabilities:
      drop:
        - ALL

policyEnforcementMode: default
policyAuditMode: false
policyCIDRMatchMode: nodes

k8sClientRateLimit:
  qps: 45
  burst: 60

cgroup:
  autoMount:
    enabled: false
  hostRoot: /sys/fs/cgroup

securityContext:
  # -- User to run the pod with
  runAsUser: 0
  # -- Run the pod with elevated privileges
  privileged: false

  capabilities:
    # -- Capabilities for the `mount-cgroup` init container
    mountCgroup:
      # Only used for 'mount' cgroup
      - SYS_ADMIN
      # Used for nsenter
      - SYS_CHROOT
      - SYS_PTRACE
    # -- capabilities for the `apply-sysctl-overwrites` init container
    applySysctlOverwrites:
      # Required in order to access host's /etc/sysctl.d dir
      - SYS_ADMIN
      # Used for nsenter
      - SYS_CHROOT
      - SYS_PTRACE
    # -- Capabilities for the `cilium-agent` container
    ciliumAgent:
      # Use to set socket permission
      - CHOWN
      # Used to terminate envoy child process
      - KILL
      # Used since cilium modifies routing tables, etc...
      - NET_ADMIN
      # Used since cilium creates raw sockets, etc...
      - NET_RAW
      # Used since cilium monitor uses mmap
      - IPC_LOCK
      # We need it for now but might not need it for >= 5.11 specially
      # for the 'SYS_RESOURCE'.
      # In >= 5.8 there's already BPF and PERMON capabilities
      - SYS_ADMIN
      # Could be an alternative for the SYS_ADMIN for the RLIMIT_NPROC
      - SYS_RESOURCE
      # Both PERFMON and BPF requires kernel 5.8, container runtime
      # cri-o >= v1.22.0 or containerd >= v1.5.0.
      # If available, SYS_ADMIN can be removed.
      - PERFMON
      - BPF
      # Allow discretionary access control (e.g. required for package installation)
      - DAC_OVERRIDE
      # Allow to set Access Control Lists (ACLs) on arbitrary files (e.g. required for package installation)
      - FOWNER
      # Allow to execute program that changes GID (e.g. required for package installation)
      - SETGID
      # Allow to execute program that changes UID (e.g. required for package installation)
      - SETUID
    # -- Capabilities for the `clean-cilium-state` init container
    cleanCiliumState:
      # Most of the capabilities here are the same ones used in the
      # cilium-agent's container because this container can be used to
      # uninstall all Cilium resources, and therefore it is likely that
      # will need the same capabilities.
      # Used since cilium modifies routing tables, etc...
      - NET_ADMIN
      # We need it for now but might not need it for >= 5.11 specially
      # for the 'SYS_RESOURCE'.
      # In >= 5.8 there's already BPF and PERMON capabilities
      - SYS_ADMIN
      # Could be an alternative for the SYS_ADMIN for the RLIMIT_NPROC
      - SYS_RESOURCE
      # Both PERFMON and BPF requires kernel 5.8, container runtime
      # cri-o >= v1.22.0 or containerd >= v1.5.0.
      # If available, SYS_ADMIN can be removed.
      - PERFMON
      - BPF
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by ekarlso** [comment]
I narrowed the culprit down to this commit: https://github.com/cilium/cilium/commit/3951a179711fed9519414d7f95eff73563180d56

**Comment by HenkVanMaanen** [comment]
Same problem but on k8s v1.29.0

**Comment by julianwiedmann** [comment]
fwiw, I [ran](https://github.com/cilium/cilium/pull/30252) your config through the grinder (https://github.com/cilium/cilium/actions/runs/7530507052/job/20497309161), and it's nowhere close to the limit:

`    verifier_test.go:254: tail_nodeport_nat_ingress_ipv4: processed 264731 insns (limit 1000000) max_states_per_insn 28 total_states 16753 peak_states 1840 mark_read 47 (2280 unverified insns)`

So either our test infrastructure is still not accurate, or something is fishy.

**Comment by ekarlso** [comment]
@julianwiedmann Can OS factors or cilium policies affect the  tail_nodeport_nat_ingress_ipv4 ?

**Comment by ekarlso** [comment]
So if I remove bpq.masquerade it works fine btw.

**Comment by HenkVanMaanen** [comment]
> @HenkVanMaanen Are you running Talos 1.6.x ?

Yes v1.6.1

**Comment by HenkVanMaanen** [comment]
> So if I remove `bpq.masquerade` it works fine btw.

This works for me too. I no longer get BPF too large errors when masquerade is false

**Comment by ekarlso** [comment]
@HenkVanMaanen I tested removing bpf.masquerade but then I am left with that return traffic back from CoreDNS times out to the DNS proxy.

**Comment by ekarlso** [comment]
@HenkVanMaanen is your cluster functional when you do remove bpf.masquerade though?

**Comment by HenkVanMaanen** [comment]
Yes it's functional. As in, I have a "who am I" pod running that I can access via: browser -> LB -> Cilium Ingress -> Service -> Deployment -> "who am I" pods.

**Comment by ekarlso** [comment]
Hi, I tested the image 5c768a4eddbb71662dc8573c0e63f910a64477df, it started fine!

**Comment by joestringer** [comment]
As far as I understand, we anticipate that #30442 should solve this issue, we're just awaiting a backport for it to the v1.15 branch.

**Comment by julianwiedmann** [comment]
https://github.com/cilium/cilium/pull/30529 landed, so I think we're good :tada: .

**Comment by joestringer** [comment]
Yep, the related fixes are in the tree for v1.15.0. Thanks everyone for collaborating on the report, debugging, fixes. Closing this out.
